### PR TITLE
Transformer

### PR DESCRIPTION
**Pull request by varisd**
- no beam search support yet
- comparison benchmark still in progress

### CODE REVIEW

File: neuralmonkey/attention/scaled_dot_product.py [hunk 1]
```diff
@@ -23,6 +24,143 @@
...
+
+
+def attention(
+        queries: tf.Tensor,
+        keys: tf.Tensor,
+        values: tf.Tensor,
+        keys_mask: tf.Tensor,
+        num_heads: int,
+        dropout_callback: Callable[[tf.Tensor], tf.Tensor],
+        masked: bool = False) -> tf.Tensor:
```
Comment (jindrahelcl) [inline neuralmonkey/attention/scaled_dot_product.py]: Docstring:
Run multi-head scaled dot-product attention.

do ``Args:`` a ``Returns:`` přesunout ty shapy co jsou v komentářích o pár řádků níž, plus:
``keys_mask`` a float Tensor for masking sequences in keys
``num_heads`` how many attention heads to use
``dropout_callback`` a function that performs dropout (so that the `train_mode` and `dropout_keep_prob` can be "closured" in)
``masked`` whether to apply triangular mask to the queries (handy in autoregressive decoders, used to mask-out the future states)

File: neuralmonkey/attention/scaled_dot_product.py [hunk 2]
```diff
@@ -23,6 +24,143 @@
...
+    weights_all = weights_4d * tf.expand_dims(tf.expand_dims(mask, 1), 1)
+
+    # normalization along time(k)
+    # norm shape: batch, head, time(q), 1
+    norm = tf.reduce_sum(weights_all, 3, keep_dims=True) + 1e-8
+    return weights_all / norm
+
+
+def mask_future(energies: tf.Tensor) -> tf.Tensor:
+    """Mask energies of keys to the right of the query."""
```
Comment (jindrahelcl) [inline neuralmonkey/attention/scaled_dot_product.py]: docstring přidat args (ener - tensor to mask) a returns (the masked tensor)

File: neuralmonkey/attention/scaled_dot_product.py [hunk 3]
```diff
@@ -23,6 +24,143 @@
...
+    """
+    x_shape = tf.shape(x)
+    x_4d = tf.reshape(tf.expand_dims(x, 2),
+                      [x_shape[0], x_shape[1], n_heads, head_dim])
+
+    return tf.transpose(x_4d, perm=[0, 2, 1, 3])
+
+
+def mask_weights(weights_4d: tf.Tensor, mask: tf.Tensor) -> tf.Tensor:
+    """Apply mask to softmax weights and renormalize."""
```
Comment (jindrahelcl) [inline neuralmonkey/attention/scaled_dot_product.py]: docstring (args weights_4d: softmax weights to renormalize + přesunout shape z komentáře, mask: float Tensor of zeros and ones, specifies valid positions in the weight tensor + přesunout shape z komentáře nííže.
returns: Renormalized distribution over valid positions, has the same shape as weights_4d

File: neuralmonkey/attention/scaled_dot_product.py [hunk 4]
```diff
@@ -23,6 +24,143 @@
 # pylint: enable=invalid-name
 
 
+def split_for_heads(x: tf.Tensor, n_heads: int, head_dim: int) -> tf.Tensor:
+    """Split a tensor for multi-head attention.
+
+    Split last dimension of 3D vector of shape (batch, time, dim) and return
+    a 4D vector with shape (batch, n_heads, time, dim/n_heads)
+    """
```
Comment (jindrahelcl) [inline neuralmonkey/attention/scaled_dot_product.py]: doplnit/přeformulovat docstring o args  a retrurns

File: neuralmonkey/attention/scaled_dot_product.py [hunk 5]
```diff
@@ -69,25 +197,23 @@ def attention(self,
...
-                    tf.layers.dense(self.attention_keys, self._head_dim,
-                                    name="keys_proj_head{}".format(i)),
-                    tf.layers.dense(self.attention_values, self._head_dim,
-                                    name="values_proj_head{}".format(i)))
-                for i in range(self.n_heads)])
-
-            context = tf.layers.dense(
-                tf.concat(head_contexts, -1), self._dimension,
-                name="output_proj")
+        # transform (batch, query_size) to (batch, 1, query_size)
```
Comment (jindrahelcl) [inline neuralmonkey/attention/scaled_dot_product.py]: convert these comments to docstring
add Args and Returns, say that this function is a API-conforming wrapper to the `attention` function defined above in the module.

File: neuralmonkey/config/utils.py [hunk 6]
```diff
@@ -28,6 +28,10 @@ def adam_optimizer(learning_rate: float = 1e-4) -> tf.train.AdamOptimizer:
     return tf.train.AdamOptimizer(learning_rate)
 
 
+def lazy_adam_optimizer(learning_rate: float = 1e-4) -> tf.train.AdamOptimizer:
+    return tf.contrib.opt.LazyAdamOptimizer(learning_rate)
```
Comment (jindrahelcl) [inline neuralmonkey/config/utils.py]: called from config -> `check_argument_types()`

Comment (jindrahelcl) [reply neuralmonkey/config/utils.py]: plus a docstring wouldn't hurt

File: neuralmonkey/decoders/transformer.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+            "word_embeddings", [len(self.vocabulary), self.dimension],
+            initializer=tf.random_uniform_initializer(-0.5, 0.5))
+
+    def embed_inputs(self, inputs: tf.Tensor) -> tf.Tensor:
+        embedded = tf.nn.embedding_lookup(self.embedding_matrix, inputs)
+        length = tf.shape(inputs)[1]
+        return embedded + position_signal(self.dimension, length)
+
+    @tensor
+    def embedded_train_inputs(self) -> tf.Tensor:
```
Comment (jlibovicky) [inline neuralmonkey/decoders/transformer.py]: Proč tohle není v předkovi?

File: neuralmonkey/decoders/transformer.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+            last_layer.temporal_states, tf.expand_dims(self.decoding_w, 0), 1,
+            "SAME")
+
+        # shape (batch, time, vocab)
+        logits = logits_unbiased + tf.reshape(self.decoding_b, [1, 1, -1])
+
+        # return logits in time-major shape
+        return tf.transpose(logits, perm=[1, 0, 2])
+
+    # @tensor
```
Comment (jlibovicky) [inline neuralmonkey/decoders/transformer.py]: Co je tohle? Vskutečnosti v předkovi?

File: neuralmonkey/decoders/transformer.py (hunk 7 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/transformer.py]: protože to nepotřebuje - tohle jsou všechny embedovaný slova v čase. v normálnim dekodéru se embeduje vždycky jen jedno v tom while loopu.

File: neuralmonkey/decoders/transformer.py (hunk 8 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/transformer.py]: alternativní implementace `train_xents` z předka s label smoothing

Comment (jindrahelcl) [reply neuralmonkey/decoders/transformer.py]: dark slide

File: neuralmonkey/attention/scaled_dot_product.py [hunk 9]
```diff
@@ -23,6 +24,143 @@
...
+    keys = split_for_heads(keys, num_heads, head_dim)
+    values = split_for_heads(values, num_heads, head_dim)
+
+    # For dot-product, we use matrix multiplication
+    # shape: batch, head, time(q), time(k) (k_channels is the matmul axis)
+    energies = tf.matmul(queries, keys, transpose_b=True)
+
+    # To protect the attention from looking ahead of time, we must
+    # replace the energies of future keys with negative infinity
+    # We use lower triangular matrix and basic tf where tricks
```
Comment (varisd) [inline neuralmonkey/attention/scaled_dot_product.py]: Ten posledni radek by snad i sel smaznout. Z te mask_future metody je to jasne.

Comment (jindrahelcl) [reply neuralmonkey/attention/scaled_dot_product.py]: Jo, zvlášť až se vyřeší jindrův review u tý funkce `mask_future`

File: neuralmonkey/attention/scaled_dot_product.py [hunk 10]
```diff
@@ -23,6 +24,143 @@
...
+    # shape: batch, head, time(q), time(k)
+    weights = tf.nn.softmax(energies)
+
+    if keys_mask is not None:
+        weights = mask_weights(weights, keys_mask)
+
+    # apply dropout to the weights (Attention Dropout)
+    weights = dropout_callback(weights)
+
+    # CODE MUSEUM
```
[the remark this answers is not in the document]
Comment (varisd) [reply neuralmonkey/attention/scaled_dot_product.py]: Nooo, not the code museum!!!

File: neuralmonkey/decoders/transformer.py (hunk 8 above)
Comment (varisd) [reply neuralmonkey/decoders/transformer.py]: takze smazat?

Comment (jlibovicky) [reply neuralmonkey/decoders/transformer.py]: Dělalo to někdy něco? Když už je ten kód napsaný, tak bych přidal někam (možná do předka?) nějaký option na potenciální label smoothing.

Comment (jindrahelcl) [reply neuralmonkey/decoders/transformer.py]: dal bych to do předka a přidal tomu option a na default to hodil vypnutý. Je ta úprava triviální? Jestli ne, tak bych to mergnul bez toho, potažmo s timhle zakomentovaným - v T2T ten label smoothing je, takže budeme-li chtít to replikovat, budeme potřebovat i tuhle featuru

Comment (varisd) [reply neuralmonkey/decoders/transformer.py]: ok, tenhle "dark" label smoothing hodim do neuralmonkey.functions (otazka je, jaky tomu dat kryci nazev), do predka pridam option label_smoothing_function, jako default bude mit default z tf.contrib.seq2seq.sequence_loss

File: neuralmonkey/decoders/transformer.py [hunk 11]
```diff
@@ -59,6 +59,8 @@ def __init__(self,
             data_id=data_id,
             max_output_len=max_output_len,
             dropout_keep_prob=dropout_keep_prob,
+            # from the "dark slide"
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/transformer.py]: tohle vyhoď a dej ten label_smoothing do parametrů initu - chceme to štelovat z konfigu

File: neuralmonkey/decoders/transformer.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+                 load_checkpoint: str = None) -> None:
+        check_argument_types()
+        AutoregressiveDecoder.__init__(
+            self,
+            name=name,
+            vocabulary=vocabulary,
+            data_id=data_id,
+            max_output_len=max_output_len,
+            dropout_keep_prob=dropout_keep_prob,
+            # from the "dark slide"
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/transformer.py]: tamto někam zapadlo tak to tady opakuju - tenhle koment vyhoď a label smoothing dej do argumentů konstruktoru - chceme to štelovat z konfigu

File: neuralmonkey/decoders/transformer.py [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+                 # TODO infer the default for these three from the encoder
+                 ff_hidden_size: int,
+                 n_heads_self: int,
+                 n_heads_enc: int,
+                 depth: int,
+                 max_output_len: int = None,
+                 dropout_keep_prob: float = 1.0,
+                 attention_dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/transformer.py]: tady by ještě měl být ten parametr na `label_smoothing`, stejně jako v ostatních potomcích `Autoregressive`, jako je třeba rekurentní dekodér.
